xarray/core/utils.py

```diff
@@ -462,12 +462,22 @@ def __repr__(self: Any) -> str:
 class ReprObject:
     """Object that prints as the given value, for use with sentinel values.
     """
+    __slots__ = ('_value', )
+
     def __init__(self, value: str):
         self._value = value
 
     def __repr__(self) -> str:
         return self._value
 
+    def __eq__(self, other) -> bool:
+        if isinstance(other, ReprObject):
+            return self._value == other._value
+        return False
+
+    def __hash__(self) -> int:
+        return hash((ReprObject, self._value))
```

Comment:
I don't think this will be stable across processes - I just tried this in two different processes:

```python

In [5]: hash(ReprObject)
Out[5]: 8774331846827

(new process)


In [2]: hash(ReprObject)
Out[2]: 8795055479397
```

Does that matter here? If so, could we change to `hash(('ReprObject', self._value))`?

Reply:
Ah, the hash is never stable across processes, I just tried the same for `'hello'`.

Does dask rely on the hash, or an equality (or something else)?

Reply:
I think it's here: `--> 708         variable = ds._variables.pop(_THIS_ARRAY)` from https://github.com/pydata/xarray/issues/3171

So then both, so this looks good - doesn't need to be consistent across processes

Reply:
xarray (actually not dask) relies on the fact that a deep copy of a key object that has gone through a round-trip on the network produces the same hash as the original. The fact that the hash changes (by design; it's an anti-DoS measure) every time you restart the interpteter is irrelevant because the ``__hash__`` method is always invoked locally. It would be, on the other hand, a grave mistake to cache it and then allow the cache to be serialised.

Reply:
To elaborate:

BAD - the hash defaults to the id, which will break as soon as the object is deep-copied, e.g. by a pickle round-trip
```python
class C:
    pass
```

GOOD:
```python
class C:
    def __hash__(self):
        return 123  # skip: actual calculation
```
BAD - it works as long as you do everything locally, but it will break as soon as you open the object in a different interpreter
```python
class C:
    _hash_cache = None

    def __hash__(self):
        if self._hash_cache is None:
            self._hash_cache = 123  # skip: actual, CPU-intensive, calculation
        return self._hash_cache
```
GOOD AGAIN:
```python
class C:
    _hash_cache = None

    def __hash__(self):
        if self._hash_cache is None:
            self._hash_cache = 123  # skip: actual, CPU-intensive, calculation
        return self._hash_cache

    def __getstate__(self):
        state = self.__dict__.copy()
        state.pop("_hash_cache", None)
        return state
```

Reply:
Thanks for the explanation! Out of interest, why is this the case: 

> BAD - it works as long as you do everything locally, but it will break as soon as you pickle the object and unpickle it in a different interpreter. 

...assuming...

>The fact that the hash changes every time you restart the interpteter is irrelevant because the __hash__ method is always invoked locally.

Reply:
I'm a little surprised that these default values end up in pickles, but I guess it can happen.

Reply:
I'm doubtful the default values would - an attribute on the class wouldn't be in the instance dict until it's assigned to